crates/flowctl/src/raw/capture.rs

```diff
@@ -0,0 +1,254 @@
+            *explicit_ack = opened.explicit_acknowledgements;
+        }
+
+        if let Some(captured) = item.captured {
+            let doc = captured.doc_json;
+
+            let sender_mutex = &channels[captured.binding as usize];
+            let mut sender = sender_mutex.lock().unwrap();
+            sender.try_send(Command::Combine(doc))?;
+        }
+
+        if let Some(ConnectorState { updated_json, merge_patch }) = item.checkpoint.and_then(|c| c.state) {
+            let mut cp = checkpoint.lock().unwrap();
+            let update = serde_json::from_str(&updated_json)?;
+            if merge_patch {
+                json_patch::merge(&mut cp, &update);
+            } else {
+                *cp = update;
+            }
+
+            for channel in channels.iter() {
+                let mut sender = channel.lock().unwrap();
+                sender.try_send(Command::Drain)?;
+            }
+
+            if *explicit_ack {
+                // Send acknowledge to connector
+                let mut ack_channel = req_send_arc.lock().unwrap();
+                ack_channel.try_send(Request {
+                    acknowledge: Some(request::Acknowledge {
+                        checkpoints: 1,
+                    }),
+                    ..Default::default()
+                })?;
+            }
+        }
+    }
+}
+
+pub struct State {
+    // Combiner of published documents.
+    combiner: doc::Combiner,
+
+    collection_name: String,
+}
+
+impl State {
+    pub fn combine_right(&mut self, doc_json: &str) -> anyhow::Result<()> {
+        let memtable = match &mut self.combiner {
+            doc::Combiner::Accumulator(accumulator) => accumulator.memtable()?,
+            _ => panic!("implementation error: combiner is draining, not accumulating"),
+        };
+        let alloc = memtable.alloc();
+
+        let mut deser = serde_json::Deserializer::from_str(doc_json);
+        let doc = doc::HeapNode::from_serde(&mut deser, alloc).with_context(|| {
+            format!(
+                "couldn't parse published document as JSON: {}",
+                doc_json
+            )
+        })?;
+
+        memtable.add(doc, false)?;
+
+        Ok(())
+    }
+
+    pub fn drain_chunk(
+        mut self,
+        out: &mut Vec<String>,
+    ) -> Result<Self, doc::combine::Error> {
+        let mut drainer = match self.combiner {
+            doc::Combiner::Accumulator(accumulator) => accumulator.into_drainer()?,
+            doc::Combiner::Drainer(d) => d,
+        };
+        let _more = drainer.drain_while(|doc, _fully_reduced| {
+            let doc_json = serde_json::to_string(&doc).expect("document serialization cannot fail");
+            out.push(doc_json);
+
+            Ok::<bool, doc::combine::Error>(out.len() != out.capacity())
+        })?;
+
+        // FIXME: do we need this? do we need to have a specific capacity for the output? At the
+        // moment I'm using an open-ended Vec
+        /*if more {
```

Comment:
You do need this test. Right now you're draining a small number of documents (probably up to eight), and then discarding the rest of the content in the combiner when you switch it back into accumulation mode.